### PR TITLE
Refactor: logic to get the rule's error message

### PR DESCRIPTION
**Pull request by jspreddy**
Refactored the logic for getting an error message related to a rule.
Don't want to have the same logic in two places.

priority wise:
1. check if a custom message is defined through validation config row.
2. check if a custom message has been set using the set_message() function
3. check if we have an error message in lang file
4. DEPRECATED support for non-prefixed keys, in lang file again
5. error message not found -> default to `"form_validation_error_message_not_set"`

### CODE REVIEW

**Comment by narfbg** [comment]
It's not the same logic.

**Comment by jspreddy** [comment]
@narfbg I understand that you are busy but can you help me with this? 
The only difference I could find is the difference in the default message used if error message line was not set anywhere in the system.

Here we are setting the line to a custom string message.
refactor__logic_to_get_the_rule_s_error_message_by_jspreddy_ _pull_request__4342_ _bcit-ci_codeigniter

Here we are using a similar message from the lang file.
refactor__logic_to_get_the_rule_s_error_message_by_jspreddy_ _pull_request__4342_ _bcit-ci_codeigniter



I believe the rest of the logic is the same as the code that I refactored from.

## The new extracted function
refactor__logic_to_get_the_rule_s_error_message_by_jspreddy_ _pull_request__4342_ _bcit-ci_codeigniter

## Block1 (823 - 840)
refactor__logic_to_get_the_rule_s_error_message_by_jspreddy_ _pull_request__4342_ _bcit-ci_codeigniter

If we remove the `!` and interchange the code inside else and elseif blocks, we get _Block2_
refactor__logic_to_get_the_rule_s_error_message_by_jspreddy_ _pull_request__4342_ _bcit-ci_codeigniter


## Block2 (640 - 655)
refactor__logic_to_get_the_rule_s_error_message_by_jspreddy_ _pull_request__4342_ _bcit-ci_codeigniter

**Comment by narfbg** [comment]
My bad, I only looked at the removed code and it's not obviously the same logic, so I closed it. Reopening and I'll leave inline comments for the stuff that you need to update.

However, unless you absolutely have to, please don't use screenshots in the future ... the steps numbering and arrows did help, but if I had to copy any of that code for debugging purposes, I wouldn't be able to.

File: system/libraries/Form_validation.php [hunk 1]
```diff
@@ -865,6 +837,47 @@ protected function _execute($row, $rules, $postdata = NULL, $cycles = 0)
 	// --------------------------------------------------------------------
 
 	/**
+	 * Get the error message for the rule
+	 *
+	 * @param 	string the rule name.
```
Comment (narfbg) [inline system/libraries/Form_validation.php]: The format is:

    @param	type	$variable_name	description

... and descriptions start with capital letters, no period at EOL

File: system/libraries/Form_validation.php [hunk 2]
```diff
@@ -865,6 +837,47 @@ protected function _execute($row, $rules, $postdata = NULL, $cycles = 0)
...
+	 * @param 	array
+	 * @return 	string
+	 */
+	private function _get_raw_error_message($rule_name, $row)
```
Comment (narfbg) [inline system/libraries/Form_validation.php]: We don't use private methods; use protected, so that it is extendable.

Also:

 - it should be named just `_get_error_message()`
 - `$rule_name` should be just `$rule` (who says it's a name? it may be a callable)
 - `$row` doesn't make sense in this context; it should be `$field`

File: system/libraries/Form_validation.php [hunk 3]
```diff
@@ -865,6 +837,47 @@ protected function _execute($row, $rules, $postdata = NULL, $cycles = 0)
...
+	{
+		$error_message = '';
```
Comment (narfbg) [inline system/libraries/Form_validation.php]: This is unnecessary, you set the variable name at all times anyway.

On another note, instead of setting it only to return it after the if block, you could just return straight from inside the conditions (and you wouldn't need the `else` in that case).

File: system/libraries/Form_validation.php [hunk 4]
```diff
@@ -865,6 +837,47 @@ protected function _execute($row, $rules, $postdata = NULL, $cycles = 0)
...
+		if (isset($this->_field_data[$row['field']]['errors'][$rule_name]))
+		{
+			$error_message = $this->_field_data[$row['field']]['errors'][$rule_name];
+		}
+		// check if a custom message has been set using the set_message() function
+		elseif (isset($this->_error_messages[$rule_name]))
+		{
+			$error_message = $this->_error_messages[$rule_name];
+		}
+		// check if we have an error message in lang file
```
Comment (narfbg) [inline system/libraries/Form_validation.php]: This comment line is completely redundant, the code is really obvious.

File: system/libraries/Form_validation.php [hunk 5]
```diff
@@ -865,6 +837,47 @@ protected function _execute($row, $rules, $postdata = NULL, $cycles = 0)
...
+		elseif (FALSE !== ($line_tmp = $this->CI->lang->line('form_validation_'.$rule_name)))
```
Comment (narfbg) [inline system/libraries/Form_validation.php]: ... it's also obvious that it's a temporary variable, why did you remove it to `$line_tmp`?

File: system/libraries/Form_validation.php [hunk 6]
```diff
@@ -865,6 +837,47 @@ protected function _execute($row, $rules, $postdata = NULL, $cycles = 0)
...
+		{
+			$error_message = $line_tmp;
+		}
+		// DEPRECATED support for non-prefixed keys, lang file again
+		elseif (FALSE !== ($line_tmp = $this->CI->lang->line($rule_name, FALSE)))
+		{
+			$error_message = $line_tmp;
+		}
+		// error message not found
```
Comment (narfbg) [inline system/libraries/Form_validation.php]: Another redundant comment.

**Comment by narfbg** [comment]
On another note ... 3.1 won't support PHP 5.2, so a closure (instead of a hard-declared method) would be just fine here and probably more performant.

File: system/libraries/Form_validation.php [hunk 7]
```diff
@@ -865,6 +837,41 @@ protected function _execute($row, $rules, $postdata = NULL, $cycles = 0)
...
+		if (isset($this->_field_data[$field]['errors'][$rule]))
+		{
+			return $this->_field_data[$field]['errors'][$rule];
+		}
+		// check if a custom message has been set using the set_message() function
+		elseif (isset($this->_error_messages[$rule]))
+		{
+			return $this->_error_messages[$rule];
+		}
+		elseif (FALSE !== ($tmp = $this->CI->lang->line('form_validation_' . $rule)))
```
Comment (narfbg) [inline system/libraries/Form_validation.php]: https://github.com/jspreddy/CodeIgniter/commit/dd149da93a2064937e490977343d152207b73815#commitcomment-15986814 was actually meant for here ... sorry.

File: system/libraries/Form_validation.php [hunk 8]
```diff
@@ -865,6 +837,41 @@ protected function _execute($row, $rules, $postdata = NULL, $cycles = 0)
 	// --------------------------------------------------------------------
 
 	/**
+	 * Get the error message for the rule
+	 *
+	 * @param 	string $rule 	The rule name
+	 * @param 	string $field
```
Comment (narfbg) [inline system/libraries/Form_validation.php]: Missing description here.

File: system/libraries/Form_validation.php [hunk 9]
```diff
@@ -865,6 +837,41 @@ protected function _execute($row, $rules, $postdata = NULL, $cycles = 0)
...
+	 *
```
Comment (narfbg) [inline system/libraries/Form_validation.php]: ... and we don't separate phpdoc stuff with empty lines. :)

File: system/libraries/Form_validation.php [hunk 10]
```diff
@@ -637,21 +637,7 @@ protected function _execute($row, $rules, $postdata = NULL, $cycles = 0)
...
-				{
-					$line = $this->_error_messages[$type];
-				}
-				elseif (FALSE === ($line = $this->CI->lang->line('form_validation_'.$type))
-					// DEPRECATED support for non-prefixed keys
-					&& FALSE === ($line = $this->CI->lang->line($type, FALSE)))
-				{
-					$line = 'The field was not set';
-				}
+				$line = $this->_get_error_message($type, $row["field"]);
```
Comment (narfbg) [inline system/libraries/Form_validation.php]: This is irrelevant, but we generally use single quotes unless doubles are necessary.

I don't really mind, but since you have to make other adjustments, you may as well change this line and 811, for consistency.

**Comment by narfbg** [comment]
I notice that you didn't leave me a comment after your latest changes. Please do that next time as commits don't issue notifications ...

... unless you write `@somebody` in the commit message, which you did, but shouldn't. Now each time somebody merges that commit in another repo I'll get spammed.

File: system/libraries/Form_validation.php (hunk 8 above)
Comment (jspreddy) [inline system/libraries/Form_validation.php]: Could the description be "The field name"

File: system/libraries/Form_validation.php (hunk 10 above)
Comment (jspreddy) [inline system/libraries/Form_validation.php]: good call. Will keep that in mind.

**Comment by jspreddy** [comment]
@narfbg Made the changes.
